rust/src/storage/s3/mod.rs

```diff
@@ -415,70 +449,77 @@ pub trait LockClient: Send + Sync + Debug {
     async fn release_lock(&self, lock: &LockItem) -> Result<bool, StorageError>;
 }
 
-#[allow(clippy::unnecessary_wraps)]
-fn try_create_lock_client(_region: Region) -> Result<Option<Box<dyn LockClient>>, StorageError> {
-    match std::env::var("AWS_S3_LOCKING_PROVIDER") {
-        Ok(p) if p.to_lowercase() == "dynamodb" => {
-            cfg_if::cfg_if! {
-                if #[cfg(feature = "dynamodb")] {
-                    let client = dynamodb_lock::DynamoDbLockClient::new(
-                        rusoto_dynamodb::DynamoDbClient::new(_region),
-                        dynamodb_lock::Options::default()
-                    );
-                    Ok(Some(Box::new(client)))
-                } else {
-                    Err(StorageError::Generic("dynamodb feature is not enabled".to_string()))
-                }
+const DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS: u32 = 10_000;
+
+impl dyn LockClient {
+    async fn rename_with_lock(
+        &self,
+        s3: &S3StorageBackend,
+        src: &str,
+        dst: &str,
+    ) -> Result<(), StorageError> {
+        let lock = self.acquire_lock_loop(src, dst).await?;
+
+        if let Some(ref data) = lock.data {
+            let data: LockData = serde_json::from_str(data)
+                .map_err(|_| StorageError::S3Generic("Lock data deserialize error".to_string()))?;
+
+            let rename_result = s3.unsafe_rename_obj(&data.source, &data.destination).await;
```

Comment:
Looks like there is no validation on lock data here, @mosyp just in case you missed the discussion in https://github.com/delta-io/delta-rs/discussions/89#discussioncomment-704821 :)

Reply:
Yes, I've definitely missed this one. That's a good catch, however I recall we specifically stated that the VersionsExists should be returned, let me have my check

Reply:
Git this, there was a case when it's required (more in discussion thread), but as it turns out that calling intended `rename` again would cover this

Reply:
@houqp @xianwill We have several choices here
Once the commit is repaired we
1) Release a lock, start acquire_loop again to get a lock with `data` we want, call rename, release lock (mostly reuse code)
2) Update a lock with new data, call rename, release lock (we have to create new update_lock function to update lock in dynamo (there's none, only put, get & release; and also we have to make a new rvn for a lock, so the lease_duration is extended, this is somewhat a _heartbeat_ from reference impl) 

We need to have updated data before calling the intended rename, otherwise, if it fails we'll most likely will break a log or lost it

Reply:
Also the problem with option 1 is that someone else might try to steal a lock and increase version etc, so it brings ambiguity to the situation

Reply:
i like opt 2 - it seems much cleaner than opt 1 and the design you describe sounds robust and has a clear implementation path.